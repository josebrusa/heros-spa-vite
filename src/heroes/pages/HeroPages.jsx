export const HeroPages = () => {
    return <h1>heroes!!!</h1>;
};
