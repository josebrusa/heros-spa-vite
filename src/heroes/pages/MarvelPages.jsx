export const MarvelPages = () => {
    return <div>MarvelPages</div>;
};
