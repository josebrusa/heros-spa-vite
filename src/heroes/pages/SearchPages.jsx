import { useNavigate, useLocation } from "react-router-dom";
import queryString from "query-string";

import { useForm } from "../hooks/useForm";

import { getHerosByName } from "../helpers";
import { HeroCard } from "../components";

export const SearchPages = () => {
    const navigate = useNavigate();
    const location = useLocation();

    const { q = "" } = queryString.parse(location.search);
    const heroes = getHerosByName(q);

    const showSearch = q.length === 0;
    const showSearchError = q.length > 0 && heroes.length === 0;

    const { searchText, onInputChange } = useForm({
        searchText: q,
    });

    const onSearchSubmit = (event) => {
        event.preventDefault();
        // if (searchText.trim().length <= 1) return;

        navigate(`?q=${searchText}`);
    };
    return (
        <>
            <div className="row">
                <h4>Search</h4>
                <hr />
                <div className="col-5">
                    <h4>Encuentra tu Heroe</h4>
                    <hr />
                    <form onSubmit={onSearchSubmit} aria-label="form">
                        <input
                            type="text"
                            placeholder="search hero"
                            className="form-control"
                            name="searchText"
                            autoComplete="off"
                            value={searchText}
                            onChange={onInputChange}
                        />
                        <button className="btn btn-outline-primary mt-3">
                            Buscar
                        </button>
                    </form>
                </div>
                <div className="col-7">
                    <h4>Heroe</h4>
                    <hr />
                    {/* {q === "" ? ( <div className="alert alert-primary"> Encuentra un Heroe </div>) : (heroes.length === 0 && (<div className="alert alert-danger"> No hay resultados del heroe {q}</div> ) )} */}
                    <div
                        className="alert alert-primary animate__animated animate__fadeIn"
                        style={{ display: showSearch ? "" : "none" }}
                    >
                        Encuentra un Heroe
                    </div>
                    <div
                        aria-label="alert-danger"
                        className="alert alert-danger animate__animated animate__fadeInRight"
                        style={{ display: showSearchError ? "" : "none" }}
                    >
                        No hay resultados del heroe {q}
                    </div>
                    <div className="d-flex flex-wrap">
                        {heroes.map((hero) => (
                            <HeroCard key={hero.id} {...hero} />
                        ))}
                    </div>
                </div>
            </div>
        </>
    );
};
