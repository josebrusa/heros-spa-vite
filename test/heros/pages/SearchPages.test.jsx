import { render, screen, fireEvent } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { SearchPages } from "../../../src/heroes/pages/SearchPages";

const mockedUseNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useNavigate: () => mockedUseNavigate,
}));

describe("Prueba en <SearchPages/>", () => {
    beforeEach(() => jest.clearAllMocks());

    test("debe de mostrar correctamente los valores por defecto", () => {
        const { container } = render(
            <MemoryRouter>
                <SearchPages />
            </MemoryRouter>
        );
        expect(container).toMatchSnapshot();
    });

    test("debe de mostrar correctamente el input y el valor del queryString", () => {
        render(
            <MemoryRouter initialEntries={["/search?q=batman"]}>
                <SearchPages />
            </MemoryRouter>
        );
        const input = screen.getByRole("textbox");
        expect(input.value).toBe("batman");

        const img = screen.getByRole("img");
        expect(img.src).toContain("/assets/heros/dc-batman.jpg");

        const alertDanger = screen.getByLabelText("alert-danger");
        expect(alertDanger.style.display).toBe("none");
    });

    test("debe de mostrar un error si no se encuentra el hero (batman 123)", () => {
        render(
            <MemoryRouter initialEntries={["/search?q=batman123"]}>
                <SearchPages />
            </MemoryRouter>
        );

        const alertDanger = screen.getByLabelText("alert-danger");
        expect(alertDanger.style.display).toBe("");
    });

    test("debe de llamar el navigate a la pantalla nueva", () => {
        const inputValue = "superman";
        render(
            <MemoryRouter initialEntries={["/search"]}>
                <SearchPages />
            </MemoryRouter>
        );

        const input = screen.getByRole("textbox");
        fireEvent.change(input, {
            target: { name: "searchText", value: inputValue },
        });

        const form = screen.getByRole("form");
        fireEvent.submit(form);

        expect(mockedUseNavigate).toHaveBeenCalledWith(`?q=${inputValue}`);
    });
});
