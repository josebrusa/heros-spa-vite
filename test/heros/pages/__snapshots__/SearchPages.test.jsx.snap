// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en <SearchPages/> debe de mostrar correctamente los valores por defecto 1`] = `
<div>
  <div
    class="row"
  >
    <h4>
      Search
    </h4>
    <hr />
    <div
      class="col-5"
    >
      <h4>
        Encuentra tu Heroe
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="search hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-3"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Heroe
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Encuentra un Heroe
      </div>
      <div
        aria-label="alert-danger"
        class="alert alert-danger animate__animated animate__fadeInRight"
        style="display: none;"
      >
        No hay resultados del heroe 
      </div>
      <div
        class="d-flex flex-wrap"
      />
    </div>
  </div>
</div>
`;
